sklearn/metrics/_regression.py

```diff
@@ -855,9 +870,17 @@ def median_absolute_error(
 
 
 def _assemble_r2_explained_variance(
-    numerator, denominator, n_outputs, multioutput, force_finite
+    numerator, denominator, n_outputs, multioutput, force_finite, xp=None
 ):
     """Common part used by explained variance score and :math:`R^2` score."""
+    if xp is None:
+        input_arrays = [numerator, denominator]
+        if multioutput is not None and not isinstance(multioutput, str):
+            input_arrays.append(multioutput)
+        xp, _ = get_namespace(*input_arrays)
+
+    device_ = device(numerator)
```

Comment:
At this point, I was a bit surprised that we did not check that numerator and denominator are on the same device by calling `device(input_arrays)`.

I think that we do that in `r2_score` so we want to avoid it then. It might be then worth to have a small comment since this is not explicit at this stage of code.

Reply:
Alternatively, we could pass the device inspected in the caller as an input argument to `_assemble_r2_explained_variance` along with `xp` and we could make both mandatory arguments to simplify the code.

Reply:
I didn't think about that here, because unlike other places where `device` is duely used with all arrays that should be consistently on the same device, this is a private function meant to be used internally only, it let the responsibility to do those checks to the caller so to avoid repeating checks that have been done already upstream (and in practice, IIRC it's indeed OK in this pr).